File: clients/java/src/main/java/dev/ricardoleal20/chopflow/examples/EchoWorker.java
package dev.ricardoleal20.chopflow.examples;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import dev.ricardoleal20.chopflow.worker.ChopFlowWorker;
import dev.ricardoleal20.chopflow.worker.ChopTask;

/**
 * Minimal Java worker. Demonstrates both registration styles: an annotated handler
 * discovered via {@code scan(this)}, and an imperative lambda.
 *
 * <p>Run against a local broker (gRPC on :8000):
 *
 * <pre>
 *   mvn -q exec:java -Dexec.mainClass=dev.ricardoleal20.chopflow.examples.EchoWorker
 * </pre>
 *
 * Then submit work from {@link ProducerExample} (or the CLI / dashboard).
 */
public final class EchoWorker {

  @ChopTask("echo")
  public JsonElement echo(JsonElement payload) {
    JsonObject out = new JsonObject();
    out.addProperty("status", "ok");
    out.add("echo", payload);
    return out;
  }

  public static void main(String[] args) {
    String broker = args.length > 0 ? args[0] : "localhost:8000";

    ChopFlowWorker worker =
        ChopFlowWorker.builder().broker(broker).tags("default").resources("cpu", 1).build();

    EchoWorker instance = new EchoWorker();
    worker.scan(instance); // discovers @ChopTask("echo")
    worker.register("ping", p -> { // explicit lambda registration
          JsonObject out = new JsonObject();
          out.addProperty("status", "ok");
          out.addProperty("pong", System.currentTimeMillis());
          return out;
        });

    System.out.println("Starting Java echo worker against " + broker + " — Ctrl+C to stop.");
    worker.startAndAwait();
  }
}

File: clients/python/examples/echo_worker.py
"""Echo worker example: register an ``echo`` handler and run until Ctrl+C."""

from chopflow import ChopFlowWorker, task


@task("echo")
def echo(payload):
    return {"status": "ok", "echo": payload}


def main() -> None:
    import sys

    broker = sys.argv[1] if len(sys.argv) > 1 else "localhost:8000"
    worker = (
        ChopFlowWorker.builder()
        .broker(broker)
        .tags("default")
        .resources("cpu", 1)
        .build()
    )
    # `echo` was registered via the @task decorator above. Also register an
    # explicit `ping` handler to show imperative registration.
    worker.register("ping", lambda p: {"status": "ok", "pong": int(__import__("time").time())})

    print(f"Starting Python echo worker against {broker} — Ctrl+C to stop.")
    worker.start_and_await()


if __name__ == "__main__":
    main()
